File: Leetcode2023/April2023/Python/MinimizeMaximumOfArray.py
class Solution:
    def minimizeArrayValue(self, nums: List[int]) -> int:
        start = 1
        end = -sys.maxsize - 1
        
        for it in nums:
            end = max(end, it)
        
        ans = 0
        
        while start <= end:
            mid = start + (end - start) // 2
            isPossible = self.helper(nums, mid)
            
            if isPossible:
                ans = mid
                end = mid - 1
            else:
                start = mid + 1
        
        return ans
    
    def helper(self, nums: List[int], expectedMax: int) -> bool:
        n = len(nums)
        arr = nums.copy()
        
        for i in range(n - 1):
            if arr[i] > expectedMax:
                return False
            
            buffer = expectedMax - arr[i]
            arr[i + 1] -= buffer
        
        return arr[n - 1] <= expectedMax

File: Leetcode2023/April2023/Java/MinimizeMaximumOfArray.java
class Solution {
    public int minimizeArrayValue(int[] nums) {
        int start = 1;
        int end = Integer.MIN_VALUE;
        for (int it : nums) {
            end = Math.max(end, it);   
        }
        
        int ans = 0;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            boolean isPossible = helper(nums, mid);
            if (isPossible) {
                ans = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        
        return ans;
    }
    
    private boolean helper(int[] nums, int expectedMax) {
        int n = nums.length;
        long[] arr = new long[n];
        for (int i = 0; i < n; i++) {
            arr[i] = nums[i];
        }
        
        for (int i = 0; i < n - 1; i++) {
            if (arr[i] > expectedMax) {
                return false;
            }
            
            long buffer = expectedMax - arr[i];
            arr[i + 1] -= buffer;
        }
        
        return arr[n - 1] <= expectedMax;
    }
}
